### ISSUE TITLE
v4 API

### ISSUE THREAD

**Issue by ittroll** [HristoKolev/TvDbSharper#25]
TheTVDB have launched v4 of their API. Access to this now requires a paid subscription and they have said they intend to deprecate   all previous API versions in a couple of months.   :-(

I was wondering if you are planning to update TvDbSharper to support the new v4 API.

https://app.swaggerhub.com/apis-docs/thetvdb/tvdb-api_v_4/4.0.0

**Comment by HristoKolev** [comment]
I know about this, of course. 

Some time ago I planned to create a v4 version of this library in some form or another. At the time the v4 authentication process was not documented on the swagger page or anywhere else, so nobody could actually test against the API. It was basically unusable on release and for some time afterwards.

I see now that there is a `/login` request in the swagger docs page so I plan to at least try to make it work.

There are some questions that I still don't have an answer for, but I will try to find out more.

**Comment by ittroll** [comment]
That is great news and much appreciated. Good luck with your endeavours.

**Comment by ittroll** [comment]
Looks like the ongoing issues with the v4 API means the older APIs will live on for a bit longer.

_“Due to a myriad of issues with the v4 API, we will be extending the cutoff date for all of our older APIs. The new cutoff date will now be October 1, 2021.”_

**Comment by cmh31909** [comment]
So what is the latest on converting this library to work with v4api?  I just got an email from TvDb stating eveything other than v4 would stop working on 10/2/2021.

**Comment by HristoKolev** [comment]
Last time I checked (a few months ago) the API wasn't usable yet. I'll check again.

**Comment by cmh31909** [comment]
Thanks for the quick reply.  They seem to think it is, as they have said v3 will no longer be supported as of 10/2/2021.

**Comment by cmh31909** [comment]
Here is a copy of the email I received from them this morning:

 


Hi TheTVDB Community Member,

 

As you likely know, a few months ago we released the latest version of our API, affectionately known as v4. Since then, TheTVDB team has been hard at work cranking out new features like expanded search, as well as developing additional endpoints and support tools for our developer community.

 

Given these recent improvements, we wanted to encourage all of our developers to upgrade to v4 if they haven’t already. And to help with that, we have pushed the deadline for sunsetting our legacy APIs to October 1, 2021. You will have until then to ensure your project is fully operating on the v4 API.

 


	

 


Additionally, our public GitHub is now officially open! You can submit support issues and connect directly with TheTVDB’s developers in our v4 API repo.

 



Take Me to GitHub

 


We’ve also just released a Python wrapper for the v4 API, which developers can now make use of. The wrapper is currently in beta, and can be found in our GitHub repo.

**Comment by HristoKolev** [comment]
Yeah. How do you subscribe for these emails? I haven't personally used the API for 4 years, kinda lost connection with these people, especially after they deleted the forums and the gitlab instance because people were complaining.

**Comment by cmh31909** [comment]
I had to give them an email address when I obtained an ApiKey to call their api.  They sent the email to that address.

**Comment by cmh31909** [comment]
The original email came in from:  TheTVDB <support@thetvdb.com>
When I responded to that, I got an email back from Emily Helwig <ehelwig@whipmedia.com>

**Comment by HristoKolev** [comment]
Interesting.

**Comment by ittroll** [comment]
It is rather annoying that they still haven't made the promised free subscription available to people who contribute data.

It would be great to get TvDbSharper updated for v4 though. If they are sending out emails it looks like they are really going to turn off of the legacy APIs this time.

**Comment by HristoKolev** [comment]
I'll start work in it as soon as I can. :)

**Comment by HristoKolev** [comment]
I started work on the v4 rewrite. Do we know what if we have the language header implemented yet? That's something that's in v3 but missing in v4.

**Comment by HristoKolev** [comment]
Just uploaded the v4.0.0 package. It has all endpoints but I didn't have any time for tests or documentation.
I manually tested the authentication endpoint and a few of the others.

When have the time I will:
* Write some kind of tests that run against the real api with some kind of notification logic so that I'm notified when it starts to fail.
* Gather documentations for the endpoints and include it in the package.
* Write some docs in the readme so that people know how to get started.

I need some help with feeding this with real data and seeing where it will fail.

I haven't had a use for this package in 4 years so I need someone else to report these issues to me. I'll make time to fix them.

**Comment by ittroll** [comment]
This is great news, many thanks. I am not a developer, but I work with a developer on a small community project which makes use of TvDbSharper. We’ll be sure to test this out and let you know,

**Comment by HristoKolev** [comment]
I pushed an update. Methods shouldn't throw exceptions anymore except for a few of the translation endpoints that throw on the server.

**Comment by shanemartin913** [comment]
Are there any examples for those of us who do not commonly use a REST library?  How to get a series or episodes...

**Comment by HristoKolev** [comment]
The closest thing to a list of examples right now is this: 

https://github.com/HristoKolev/TvDbSharper/blob/master/test/ManualTests/Program.cs

**Comment by neogentrics** [comment]
This issue tracks the original, large-scale effort to migrate TvDbSharper to the v4 API. The original author (HristoKolev) did the initial work to release the first v4 version.

As of late 2022, the library had become out-of-sync with the live v4 API, leading to multiple bugs (like #32, #33, #34, #36, #37, and #38).

I have now completed the work to modernize this v4 client and bring it up-to-date with the current API specification. This effort included:

Framework Modernization: The library has been upgraded from .NET Standard 1.1 to .NET Standard 2.0, and the test projects have been upgraded to .NET 8.

API Client Refresh: The GenerateDto project was repaired and run against the latest swagger.yml specification.

New Endpoints Added: All new API endpoints (like getAllEpisodes, getUserInfo, getSeriesBaseBySlug, etc.) are now included in the client.

Runtime Bug Fixes: Fixed numerous JsonSerializationException errors by adding overrides for cases where the live API (e.g., company_type) differs from the official documentation (e.g., companyType).

Data Mismatch Fixes: Corrected for new, undocumented fields in the live API (like absoluteNumber, remoteIds) and nullable properties (like averageRuntime).

Error Handling: Updated the 401 Unauthorized error parsing to handle the modern API response.

The library now builds cleanly, all manual tests pass against the live API, and the client is fully functional.

This completes the v4 migration and modernization. All fixes are included in the commits ending in 95a49d5e8426ca1a6c2365e2d76e8eda09e61bca, and this issue can be closed.
